spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala

```diff
@@ -103,23 +103,30 @@ object DistriOptimizer {
     trainSummary: Option[TrainSummary],
     validationSummary: Option[ValidationSummary],
     isOverWrite: Boolean
-  )(implicit ev: TensorNumeric[T]) = {
+  )(implicit ev: TensorNumeric[T]): Unit = {
     val sc = dataset.originRDD().sparkContext
     val partitionNum = dataset.originRDD().partitions.length
     var wallClockTime = 0L
     var lastEpochTime = 0L
-    val driverState = T("epoch" -> optimMethod.state.get[Int]("epoch").getOrElse(1),
+    val driverState = T(
+      "epoch" -> optimMethod.state.get[Int]("epoch").getOrElse(1),
       "neval" -> optimMethod.state.get[Int]("neval").getOrElse(1),
       "Loss" -> optimMethod.state.get[Float]("Loss").getOrElse(Float.PositiveInfinity),
-      "score" -> optimMethod.state.get[Float]("score").getOrElse(0f))
-    val _subModelNumber = Engine.getEngineType match {
+      "score" -> optimMethod.state.get[Float]("score").getOrElse(0f)
+    )
+    val _subModelNumber = Engine.getEngineType() match {
       case MklBlas => coresPerNode
-      case _ => throw new IllegalArgumentException()
     }
-    var accumulateCount = 0
+
+    logger.info("Count dataset")
+    val countBefore = System.nanoTime()
+    val numSamples = dataset.data(train = false).map(_.size()).reduce(_ + _)
+    val countAfter = System.nanoTime()
+    logger.info(s"Count dataset complete. Time elapsed: ${(countAfter - countBefore) / 1e9}s")
+
```

Comment:
As my discussion with @yiheng, we think we should leave a warning here when `numSamples != dataset.size()`. Because
```
If the user is caching RDD[MiniBatch] directly, data in MiniBatches is fixed in each partitions, we just choose one MiniBatch from the RDD[MiniBatch]'s partition randomly. This may leads the model couldn't converge to the best possible model.
If user is caching RDD[Sample], in each iteration, we randomly transform a few sample to a MiniBatch. So data in every MiniBatch is random, this random data is better.
```